persistent/Database/Persist/Quasi/Internal/ModelParser.hs

```diff
@@ -0,0 +1,680 @@
+charLiteral :: Parser Char
+charLiteral = label "literal character" $ do
+    char1 <- anySingle
+    case char1 of
+        '\\' -> do
+            char2 <- anySingle
+            case char2 of
+                '(' -> pure '('
+                ')' -> pure ')'
+                '\\' -> pure '\\'
+                '\"' -> pure '\"'
+                '\'' -> pure '\''
+                _ -> unexpected (Tokens $ char2 :| [])
+        _ -> pure char1
+
+equality :: Parser Token
+equality = label "equality expression" $ do
+    L.lexeme spaceConsumer $ do
+        lhs <- some contentChar
+        _ <- char '='
+        rhs <-
+            choice
+                [ quotation'
+                , sqlLiteral
+                , parentheticalInner
+                , some $ contentChar <|> char '(' <|> char ')'
+                ]
+        pure $ Equality (Text.pack lhs) (Text.pack rhs)
+  where
+    parentheticalInner = do
+        str <- parenthetical'
+        pure . init . drop 1 $ str
+
+sqlTypeName :: Parser String
+sqlTypeName =
+    some $
+        choice
+            [ alphaNumChar
+            , char '_'
+            ]
+
+sqlLiteral :: Parser String
+sqlLiteral = label "SQL literal" $ do
+    quote <- L.lexeme spaceConsumer $ char '\'' *> manyTill charLiteral (char '\'')
+    st <- optional $ do
+        colons <- string "::"
+        tn <- sqlTypeName
+        pure $ colons <> tn
+    pure $
+        mconcat
+            [ "'"
+            , quote
+            , "'"
+            , fromMaybe "" st
+            ]
+
+quotation :: Parser Token
+quotation = label "quotation" $ do
+    str <- L.lexeme spaceConsumer quotation'
+    pure . Quotation $ Text.pack str
+
+quotation' :: Parser String
+quotation' = char '"' *> manyTill charLiteral (char '"')
+
+parenthetical :: Parser Token
+parenthetical = label "parenthetical" $ do
+    str <- L.lexeme spaceConsumer parenthetical'
+    pure . Parenthetical . Text.pack . init . drop 1 $ str
+
+parenthetical' :: Parser String
+parenthetical' = do
+    str <- between (char '(') (char ')') q
+    pure $ "(" ++ str ++ ")"
+  where
+    q = mconcat <$> some (c <|> parenthetical')
+    c = (: []) <$> choice [contentChar, nonLineSpaceChar, char '"']
+
+blockKey :: Parser Token
+blockKey = label "block key" $ do
+    fl <- upperChar
+    rl <- many alphaNumChar
+    pure . BlockKey . Text.pack $ fl : rl
+
+ptext :: Parser Token
+ptext = label "plain token" $ do
+    str <- L.lexeme spaceConsumer $ some contentChar
+    pure . PText . Text.pack $ str
+
+-- @since 2.16.0.0
+anyToken :: Parser Token
+anyToken =
+    choice
+        [ try equality
+        , quotation
+        , parenthetical
+        , ptext
+        ]
+
+class Block a where
+    blockFirstPos :: a -> SourcePos
+    blockMembers :: a -> [Member]
+    blockSetMembers :: [Member] -> a -> a
+    blockSetNELMembers :: NonEmpty Member -> a -> a
+    blockSetDocCommentBlock :: Maybe DocCommentBlock -> a -> a
```

Comment:
This isn't an idiomatic use of type classes. Where you reach for a class in OOP, in Haskell, you want to reach for a datatype.

```haskell
data Block = Block
    { blockFirstPos :: SourcePos
    , blockMembers :: [Member]
    , blockDocCommentBlock :: Maybe DocCommentBlock
    }
```

And then `EntityBlock` and `ExtraBlock` both contain a `Block` (along with whatever else they end up having)